src/linear_algebra/Hermitian_inner_product.lean

```diff
@@ -0,0 +1,669 @@
+begin
+  dunfold is_normalised,
+  dunfold normalise,
+  dunfold herm_norm,
+  rw [inprod_smul_left, inprod_smul_right, conj_of_real,
+    mul_re, mul_re, of_real_im, zero_mul, zero_mul,
+    sub_zero, sub_zero, of_real_re, ←sqrt_one,
+    sqrt_inj (mul_nonneg (inv_nonneg.mpr (sqrt_nonneg (x ₀ x).re)) (mul_nonneg (inv_nonneg.mpr (sqrt_nonneg (x ₀ x).re)) (inprod_self_re_nonneg x))) (zero_le_one), 
+    ←sqrt_inv, ←field.mul_assoc],
+  rw mul_self_sqrt,
+  rw field.inv_mul_cancel, 
+    exact (ne_zero_im_zero_imp_re_ne_zero (inprod_self_ne_zero_iff.mpr ho) (inprod_self_im x)),
+    exact (inv_nonneg.mpr (inprod_self_re_nonneg x)),
+end
+
+def normalise_set :
+set α → set α := image(λ x, (↑|x|⁻¹ : ℂ) • x)
+
+lemma normalise_set_normalises 
+(A : set α) (Ho : (0 : α) ∉ A) : 
+∀ x ∈ normalise_set(A), is_normalised x :=
+begin
+  dunfold normalise_set,
+  dunfold image, 
+  intros,
+  have Ha : ∃ (a : α), a ∈ A ∧ normalise a = x,
+  { rw mem_set_of_eq at H, 
+    exact H },
+  cases Ha with a Hl,
+  rw ←Hl.right,
+  have ho : a ≠ 0,
+  { intros h,
+    rw h at Hl,
+    exact Ho Hl.left },
+  exact normalise_normalises a ho,
+end
+
+def herm_ortho (x y : α) : Prop := x ₀ y = 0 
+
+notation a ⊥ b := herm_ortho a b 
+
+def herm_ortho_sym (x y : α) :
+(x ⊥ y) ↔ (y ⊥ x) := @ortho_sym ℂ α _ conj.ring_invo _ _ (to_sym_sesq_form α) _ _ 
+
+lemma ortho_refl_iff_zero (x : α) : 
+(x ⊥ x) ↔ x = 0 := inprod_self_eq_zero_iff  
+
+def herm_ortho_smul_left (x y : α) (a : ℂ) (ha : a ≠ 0) : 
+(x ⊥ y) ↔ ((a • x) ⊥ y) := @ortho_smul_left _ _ _ _ _ _ (herm_inner_product_space.to_sym_sesq_form α).to_sesq_form _ _ _ ha
+
+def herm_ortho_smul_right (x y : α) (a : ℂ) (ha : a ≠ 0) : 
+(x ⊥ y) ↔ (x ⊥ (a • y)) := @ortho_smul_right _ _ _ _ _ _ (herm_inner_product_space.to_sym_sesq_form α).to_sesq_form _ _ _ ha
+
+theorem ortho_imp_not_lindep (x y : α) (hx : x ≠ 0) (hy : y ≠ 0) : 
+(x ⊥ y) → ¬∃ (a : ℂ), (a ≠ 0) ∧ (x = a • y ∨ a • x = y) :=
+begin
+  intros H1,
+  apply not_exists.mpr,
+  intros a,
+  intros H2,
+  unfold herm_ortho at H1,
+  cases H2 with ha H2,
+  cases H2,  
+  { rw H2 at H1,
+    rw [inprod_smul_left, mul_eq_zero] at H1,
+    cases H1,
+    { trivial }, 
+
+    { exact hy ((inprod_self_eq_zero_iff).mp H1) }},
+    
+  { rw ←H2 at H1,
+    rw [inprod_smul_right, mul_eq_zero] at H1,
+    cases H1,
+    { exact ha ((conj_eq_zero).mp H1) },
+
+    { exact hx ((inprod_self_eq_zero_iff).mp H1) }}
+end 
+
+theorem pythag_iden (x y : α) (H : x ⊥ y) :
+|x + y|^2 = |x|^2 + |y|^2 :=
+begin
+  dunfold herm_norm,
+  rw [sqr_sqrt (inprod_self_re_nonneg (x + y)), sqr_sqrt (inprod_self_re_nonneg x),
+    sqr_sqrt (inprod_self_re_nonneg y), inprod_self_add],
+  unfold herm_ortho at H,
+  rw [←conj_inprod x y, H, conj_zero,
+    field.add_zero, field.add_zero, add_re],
+end
+
+def is_ortho_set (s : set α) :=
+∀ x y ∈ s, (x ⊥ y) ↔ x ≠ y 
+
+def is_orthonormal_set (s : set α) :=
+is_ortho_set s ∧ ∀ x ∈ s, is_normalised x
+
+noncomputable def proj (x y : α) :=
```

Comment:
The argument order is wrong. You want `proj x` to be the map projecting onto the line spanned by `x`.

Reply:
fixed